torch/_ops.py

```diff
@@ -218,32 +282,54 @@ def inner(fn):
                 self.python_key_mode_table[mode] = fn
                 return fn
 
-            assert isinstance(dispatch_key_or_mode, torch._C.DispatchKey)  # type: ignore[attr-defined]
+            assert isinstance(dispatch_key_or_mode, torch._C.DispatchKey)
             assert (
-                dispatch_key_or_mode != torch._C.DispatchKey.Python  # type: ignore[attr-defined]
+                dispatch_key_or_mode != torch._C.DispatchKey.Python
             ), "Please register a mode for the torch._C.DispatchKey.Python key instead."
 
             self.py_kernels[dispatch_key_or_mode] = fn
             return fn
 
         return inner
 
-    def dispatch(self, dispatch_key, *args, **kwargs):
-        if dispatch_key == torch._C.DispatchKey.Python:  # type: ignore[attr-defined]
-            # TODO(voz): We should walk all the nodes here / turn it into a list, topmode is ok for now.
-            curr_mode = type(torch._C._get_torch_dispatch_mode())
-            assert (
-                curr_mode is not None
-            ), "Illegal invocation of dispatch on torch._C.DispatchKey.Python without a mode."
-            if curr_mode not in self.python_key_mode_table:
-                return self._op_dk(dispatch_key, *args, **kwargs)
-            # TODO(voz): The idea behind this is that we do not yet support dispatch by key + mode, only key.
-            return self.python_key_mode_table[curr_mode](*args, **kwargs)
+    def __getattr__(self, attr):
+        if len(attr) == 0 or not attr[0].isupper():
+            raise AttributeError()
 
-        if dispatch_key in self.py_kernels:
-            return self.py_kernels[dispatch_key](*args, **kwargs)
-        else:
-            return self._op_dk(dispatch_key, *args, **kwargs)
+        try:
+            key = torch._C._dispatch_key_parse(attr)
+        except Exception as e:
+            import traceback
+
+            traceback.print_exc()
+            raise AttributeError()
+
+        if key == torch._C.DispatchKey.Python:
+            if not self.python_key_mode_table:
+                setattr(self, attr, key)
+                return key
+
+            def handler(*args, **kwargs):
+                # TODO: We also need to handle tensor subclasses here
+                # TODO(voz): We should walk all the nodes here / turn it into a list, topmode is ok for now.
+                curr_mode = type(torch._C._get_torch_dispatch_mode())
+                assert (
+                    curr_mode is not None
+                ), "Illegal invocation of dispatch on torch._C.DispatchKey.Python without a mode."
+                if curr_mode not in self.python_key_mode_table:
+                    # TODO: This path is slow, should generally encourage this
+                    # case to not happen
+                    return self._op_dk(key, *args, **kwargs)
```

Comment:
hmm still reading, but:

Wouldn't it be faster (and remove a case) to folder this case into here?
```
            if not self.python_key_mode_table:
                setattr(self, attr, key)
                return key
```

If I'm following correctly, it looks like when the given mode isn't registered to the PyDispatcher, we're returning a python handler that just redispatches to the python key. But it seems like the point of this PR is that we want to fast-path the case where we're just using the ordinary dispatcher, so we may as well return a dispatch key instead of a python handler when `not self.python_key_mode_table or curr_mode not in self.python_key_mode_table`.

Reply:
I'm not sure exactly what you are suggesting. But if you are proposing:

```
        if key == torch._C.DispatchKey.Python:
            curr_mode = type(torch._C._get_torch_dispatch_mode()) 
            if not self.python_key_mode_table or curr_mode not in self.python_key_mode_table:
                setattr(self, attr, key)
                return key
```

This will not work. In general, the current mode may vary on subsequent invocations, so you cannot pre-compute it at the first time the getattr is called

Reply:
ohh. That makes sense - thanks for the clarification.